[SwiftUI] Expose an async sequence for navigations instead of using Observation

rr-codes opened the pull request:
#### d2dff1e3de739285ab34ae4ff83afed9a5a4679a
<pre>
[SwiftUI] Expose an async sequence for navigations instead of using Observation
<a href="https://bugs.webkit.org/show_bug.cgi?id=294653">https://bugs.webkit.org/show_bug.cgi?id=294653</a>
<a href="https://rdar.apple.com/152414525">rdar://152414525</a>

Reviewed by Aditya Keerthi.

The current `WebPage` API has a `currentNavigationEvent` property that is intended to be used with
the Observation framework to make it possible for clients to track changes to this property. However,
after further consideration and research, this approach was discovered to be flawed for several
reasons, including:

- The `currentNavigationEvent` API property is effectively useless on its own, and is only useful
when converted to a sequence.
- The `currentNavigationEvent` property is nullable, only because it is `nil` at the very beginning,
but is never nil again once it&apos;s set, so client have to use a lot of optional chaining or force unwrapping
- One of the most common use cases of tracking navigations is to load some resource and wait for it to
be completed, which is cumbersome and difficult to do currently
- Errors are not handled idiomatically; instead of throwing errors, the API treats them as just another e
vent, which is an anti-pattern
- Task cancellation is not supported

As such, this replaces this API with one that uses async sequences directly, providing
a significantly better developer experience, and improved correctness.

This new design is advantageous for several reasons:

 - It makes it _substantially_ easier to track a specific navigation and/or wait for a specific event,
 since now it&apos;s just a trivial for-await loop
 - It supports throwing errors instead of treating them like normal values
 - No coalescing happens
 - The API is much easier to reason about and have expectations for
 - The sequence can support Task cancellation, which would result in the webpage stopping loading of its resources

* Source/WebKit/UIProcess/API/Swift/WebPage+Navigation.swift:

Expose new supporting Navigation enum types to represent navigation events and navigation errors. Note that the general
&quot;failed&quot; navigation is communicated to the client directly from the underlying NSError thrown.

* Source/WebKit/UIProcess/API/Swift/WebPage.swift:
(navigations):

Introduce a new `navigations` property, which returns a new indefinite async sequence each time it is called. The sequence
will produce events for all navigations that happen throughout the lifetime of the WebPage, unless a navigation error happens,
or the sequence&apos;s Task is cancelled.

(scopedNavigations):
(indefiniteNavigations):
(load(_:)):
(load(_:mimeType:characterEncoding:baseURL:)):
(load(_:baseURL:)):
(load(simulatedRequest:response:responseData:)):
(load(simulatedRequest:responseHTML:)):
(reload(_:)):

Update the loading APIs to return an async sequence.

(addNavigationEvent(_:any:for:)):

A helper function for the navigation delegate adapter to use which internally handles the different continuations.

(createIndefiniteNavigationSequence):
(toNavigationSequence(_:)):

Keep track of all the async stream continuations produced by the load functions so that they can later be updated and then
removed when no longer needed. There should theoretically be a maximum of two continuations in this dictionary at a time (the
case where a new load starts, cancelling the prior load). When a load sequence&apos;s Task is cancelled, the load is also cancelled.

(currentNavigationEvent): Deleted.

This is essentially the same as `scopedNavigations` but instead stores all the sequences generated by `navigations`.

* Source/WebKit/UIProcess/Cocoa/WKNavigationDelegateAdapter.swift:
(WKNavigationDelegateAdapter.yieldNavigationProgress(_:cocoaNavigation:)):
(WKNavigationDelegateAdapter.failNavigationProgress(_:cocoaNavigation:)):
(WKNavigationDelegateAdapter.webView(_:didFailProvisionalNavigation:withError:)):
(WKNavigationDelegateAdapter.webView(_:didFail:withError:)):

Fix a minor correctness bug that is exposed through the new API shape; if a navigation never gets passed starting provisionally
(such as if a URLSchemeHandler never finishes), then no navigation events will be added to any sequences because there will
be no presentation update. Since the presentation update is only actually useful for `finished` anyways, limit doing so to that case.

* Tools/SwiftBrowser/Source/ViewModel/BrowserViewModel.swift:
(BrowserViewModel.didReceiveNavigationEvent(_:)):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKit Swift/WebPageNavigationTests.swift: Added.
(NeverLoadingSchemeHandler.reply(for:)):
(WebPageNavigationTests.basicNavigationProducesExpectedNavigationEvents):
(WebPageNavigationTests.failedNavigationProducesExpectedNavigationError):
(WebPageNavigationTests.explicitlyStopLoadingProgrammaticNavigation):
(WebPageNavigationTests.stopLoadingProgrammaticNavigationViaTaskCancellation):
* Tools/TestWebKitAPI/Tests/WebKit Swift/WebPageTests.swift:

Canonical link: <a href="https://commits.webkit.org/296461@main">https://commits.webkit.org/296461@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3a65fdc0ac11b795d9d4ae4ffe388dfc84b1142a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108597 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28258 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18682 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113805 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/58987 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/110560 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28947 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36812 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82483 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111545 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/22976 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97819 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62920 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22394 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15957 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58519 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/92345 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/16008 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/116926 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35651 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/26284 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/91506 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/36024 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94089 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91310 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23259 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36208 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13972 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/31466 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35552 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/41085 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35262 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38608 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36940 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKit Swift/WebPageNavigationTests.swift:
```diff
@@ -0,0 +1,148 @@
+    }
+
+    @Test
+    func explictlyStopLoadingProgrammaticNavigation() async throws {
```
```suggestion
    func explicitlyStopLoadingProgrammaticNavigation() async throws {
```

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKit Swift/WebPageNavigationTests.swift:
```diff
@@ -0,0 +1,148 @@
+        configuration.urlSchemeHandlers[NeverLoadingSchemeHandler.scheme] = NeverLoadingSchemeHandler()
+        let page = WebPage(configuration: configuration)
+
+        let sequence = page.load(URLRequest(url: URL(string: "never-loading:///index.html")!))
```
```suggestion
        configuration.urlSchemeHandlers[NeverLoadingSchemeHandler.scheme] = NeverLoadingSchemeHandler()

        let page = WebPage(configuration: configuration)
        let sequence = page.load(URLRequest(url: URL(string: "never-loading:///index.html")!))
```

rr-codes committed:
[SwiftUI] Expose an async sequence for navigations instead of using Observation
https://bugs.webkit.org/show_bug.cgi?id=294653
rdar://152414525

Reviewed by Aditya Keerthi.

The current `WebPage` API has a `currentNavigationEvent` property that is intended to be used with
the Observation framework to make it possible for clients to track changes to this property. However,
after further consideration and research, this approach was discovered to be flawed for several
reasons, including:

- The `currentNavigationEvent` API property is effectively useless on its own, and is only useful
when converted to a sequence.
- The `currentNavigationEvent` property is nullable, only because it is `nil` at the very beginning,
but is never nil again once it's set, so client have to use a lot of optional chaining or force unwrapping
- One of the most common use cases of tracking navigations is to load some resource and wait for it to
be completed, which is cumbersome and difficult to do currently
- Errors are not handled idiomatically; instead of throwing errors, the API treats them as just another e
vent, which is an anti-pattern
- Task cancellation is not supported

As such, this replaces this API with one that uses async sequences directly, providing
a significantly better developer experience, and improved correctness.

This new design is advantageous for several reasons:

 - It makes it _substantially_ easier to track a specific navigation and/or wait for a specific event,
 since now it's just a trivial for-await loop
 - It supports throwing errors instead of treating them like normal values
 - No coalescing happens
 - The API is much easier to reason about and have expectations for
 - The sequence can support Task cancellation, which would result in the webpage stopping loading of its resources

* Source/WebKit/UIProcess/API/Swift/WebPage+Navigation.swift:

Expose new supporting Navigation enum types to represent navigation events and navigation errors. Note that the general
"failed" navigation is communicated to the client directly from the underlying NSError thrown.

* Source/WebKit/UIProcess/API/Swift/WebPage.swift:
(navigations):

Introduce a new `navigations` property, which returns a new indefinite async sequence each time it is called. The sequence
will produce events for all navigations that happen throughout the lifetime of the WebPage, unless a navigation error happens,
or the sequence's Task is cancelled.

(scopedNavigations):
(indefiniteNavigations):
(load(_:)):
(load(_:mimeType:characterEncoding:baseURL:)):
(load(_:baseURL:)):
(load(simulatedRequest:response:responseData:)):
(load(simulatedRequest:responseHTML:)):
(reload(_:)):

Update the loading APIs to return an async sequence.

(addNavigationEvent(_:any:for:)):

A helper function for the navigation delegate adapter to use which internally handles the different continuations.

(createIndefiniteNavigationSequence):
(toNavigationSequence(_:)):

Keep track of all the async stream continuations produced by the load functions so that they can later be updated and then
removed when no longer needed. There should theoretically be a maximum of two continuations in this dictionary at a time (the
case where a new load starts, cancelling the prior load). When a load sequence's Task is cancelled, the load is also cancelled.

(currentNavigationEvent): Deleted.

This is essentially the same as `scopedNavigations` but instead stores all the sequences generated by `navigations`.

* Source/WebKit/UIProcess/Cocoa/WKNavigationDelegateAdapter.swift:
(WKNavigationDelegateAdapter.yieldNavigationProgress(_:cocoaNavigation:)):
(WKNavigationDelegateAdapter.failNavigationProgress(_:cocoaNavigation:)):
(WKNavigationDelegateAdapter.webView(_:didFailProvisionalNavigation:withError:)):
(WKNavigationDelegateAdapter.webView(_:didFail:withError:)):

Fix a minor correctness bug that is exposed through the new API shape; if a navigation never gets passed starting provisionally
(such as if a URLSchemeHandler never finishes), then no navigation events will be added to any sequences because there will
be no presentation update. Since the presentation update is only actually useful for `finished` anyways, limit doing so to that case.

* Tools/SwiftBrowser/Source/ViewModel/BrowserViewModel.swift:
(BrowserViewModel.didReceiveNavigationEvent(_:)):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKit Swift/WebPageNavigationTests.swift: Added.
(NeverLoadingSchemeHandler.reply(for:)):
(WebPageNavigationTests.basicNavigationProducesExpectedNavigationEvents):
(WebPageNavigationTests.failedNavigationProducesExpectedNavigationError):
(WebPageNavigationTests.explicitlyStopLoadingProgrammaticNavigation):
(WebPageNavigationTests.stopLoadingProgrammaticNavigationViaTaskCancellation):
* Tools/TestWebKitAPI/Tests/WebKit Swift/WebPageTests.swift:

Canonical link: https://commits.webkit.org/296461@main
